core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -301,6 +303,116 @@ class TransactionStateManager(brokerId: Int,
   }
 
+  def appendTransactionToLog(transactionalId: String,
+                             txnMetadata: TransactionMetadata,
+                             responseCallback: Errors => Unit) {
+
+    // generate the message for this transaction metadata
+    val keyBytes = TransactionLog.keyToBytes(transactionalId)
+    val valueBytes = TransactionLog.valueToBytes(txnMetadata)
+    val timestamp = time.milliseconds()
+
+    val records = MemoryRecords.withRecords(TransactionLog.EnforcedCompressionType, new SimpleRecord(timestamp, keyBytes, valueBytes))
+
+    val topicPartition = new TopicPartition(Topic.TransactionStateTopicName, partitionFor(transactionalId))
+    val recordsPerPartition = Map(topicPartition -> records)
+
+    // set the callback function to update transaction status in cache after log append completed
+    def updateCacheCallback(responseStatus: collection.Map[TopicPartition, PartitionResponse]): Unit = {
+      // the append response should only contain the topics partition
+      if (responseStatus.size != 1 || !responseStatus.contains(topicPartition))
+        throw new IllegalStateException("Append status %s should only have one partition %s"
+          .format(responseStatus, topicPartition))
+
+      val status = responseStatus(topicPartition)
+
+      var responseError = if (status.error == Errors.NONE) {
+        Errors.NONE
+      } else {
+        debug(s"Transaction state update $txnMetadata for $transactionalId failed when appending to log " +
+          s"due to ${status.error.exceptionName}")
+
+        // transform the log append error code to the corresponding coordinator error code
+        status.error match {
+          case Errors.UNKNOWN_TOPIC_OR_PARTITION
+               | Errors.NOT_ENOUGH_REPLICAS
+               | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND
+               | Errors.REQUEST_TIMED_OUT => // note that for timed out request we return NOT_AVAILABLE error code to let client retry
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.COORDINATOR_NOT_AVAILABLE} to the client")
+
+            Errors.COORDINATOR_NOT_AVAILABLE
+
+          case Errors.NOT_LEADER_FOR_PARTITION =>
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.NOT_COORDINATOR} to the client")
+
+            Errors.NOT_COORDINATOR
+
+          case Errors.MESSAGE_TOO_LARGE
+               | Errors.RECORD_LIST_TOO_LARGE
+               | Errors.INVALID_FETCH_SIZE =>
+
+            error(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning UNKNOWN error code to the client")
+
+            Errors.UNKNOWN
```

Comment:
This would be a fatal error right? What would client be expected to do in this case? Should we have a more descriptive error code? 

Retries will not help in these cases, too, so it's best to be as explicit as possible IMO.